Blish HUD/Controls/TextInputBase.cs

```diff
@@ -0,0 +1,718 @@
+ *  copies or substantial portions of the Software.
+
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Blish_HUD.Input;
+using Microsoft.Xna.Framework;
+using Microsoft.Xna.Framework.Graphics;
+using MonoGame.Extended.BitmapFonts;
+using Microsoft.Xna.Framework.Input;
+
+namespace Blish_HUD.Controls {
+
+    /// <summary>
+    /// The base of controls such as the <see cref="TextBox"/> or <see cref="MultilineTextBox"/> which accept text input from the user.
+    /// </summary>
+    public abstract class TextInputBase : Control {
+
+        private static readonly Logger Logger = Logger.GetLogger<TextInputBase>();
+
+        protected static readonly char[] WordSeperators = { ' ', '\n', '`', '~', '!', '@', '#', '%', '^', '&', '*', '(', ')', '-', '=', '+', '[', '{', ']', '}', '\\', '|', ';', ':', '\'', '"', ',', '.', '<', '>', '/', '?' };
+
+        #region Load Static
+
+        private static readonly Color _highlightColor;
+
+        static TextInputBase() {
+            _highlightColor = new Color(92, 80, 103, 150);
+        }
+
+        #endregion
+
+        protected const char NEWLINE = '\n';
+
+        /// <summary>
+        /// Fires when the <see cref="Text"/> is changed.
+        /// </summary>
+        public event EventHandler<EventArgs> TextChanged;
+
+        /// <summary>
+        /// Fires when the <see cref="CursorIndex"/> is changed.
+        /// </summary>
+        public event EventHandler<ValueEventArgs<int>> CursorIndexChanged;
+
+        protected void OnTextChanged(ValueChangedEventArgs<string> e) => TextChanged?.Invoke(this, e);
+
+        protected void OnCursorIndexChanged(ValueEventArgs<int> e) {
+            _cursorMoved = true;
+
+            UpdateScrolling();
+
+            CursorIndexChanged?.Invoke(this, e);
+        }
+
+        protected string _text = string.Empty;
+
+        /// <summary>
+        /// Gets or sets the text of the control.
+        /// </summary>
+        public string Text {
+            get => _text;
+            set => SetText(value, false);
+        }
+
+        protected int _maxLength = int.MaxValue;
+
+        /// <summary>
+        /// Gets or sets the maximum character length of the control.
+        /// </summary>
+        public int MaxLength {
+            get => _maxLength;
+            set {
+                if (SetProperty(ref _maxLength, value)) {
+                    this.Text = _text.Substring(0, Math.Min(_maxLength, _text.Length));
```

Comment:
You have a property for "Text" for a reason. Don't know how i feel about it to access the backendfield directly instead using the getter through the property
(this also applies to every other use of a backendfield too)

Reply:
> You have a property for "Text" for a reason.

That reason is primarily public access from other code.  In the case of the `Text` property, to ensure that when text is set the undo/redo buffers are updated, cursor position is handled properly, invalid characters are moved, etc.

To avoid a jumble of using both backend and properties within the owning class, I've generally kept to a pattern of only using the public properties if the additional logic in the getter or setter is actually needed when accessing or setting them internally.

This is especially important with setters as all of these setters will fire an event and several will invalidate the control layout if updated using the public setter.  You can see an example of this within `DeleteSection` where we use the property/setter just for `SelectionEnd` to ensure that the control isn't invalidated until it-itself is set and doesn't get fired twice (which would happen if `CursorIndex` was used instead of `_cursorIndex`).

Reply:
Can be resolved